README
How to create a private Debian repository:

1. Run deb-add.sh to add the .deb files to /var/www/untangle
    deb-add.sh /var/www/untangle foo.deb  [more .deb files]

2. Once all .deb files are added, run deb-scan.sh to recreate the Packages.gz file
    deb-scan.sh /var/www/untangle

Repeat the steps when you want to update any .deb file.
If you add adding a new package, add an entry in override.testing.untangle.
The fields for the entry are: package name, package priority, and package section.

To use the repository (atleast this example), add this line to your sources.list file:
deb http://bebe/untangle     testing untangle



Tools:
push.sh
    sync to local repositories
    ./push.sh /var/www/dmorris/ /var/www/untangle/

push-release.sh
    pushes /var/www/untangle to release.untangle.com
    removes non public transforms

push-release-alpha.sh
    pushes /var/www/untangle to release-alpha.untangle.com
    removes non public transforms

deb-add.sh
    addes deb files ($2+) to repository ($1)

deb-scan.sh
    Scans debian packages in repository ($1) and builds Packages list

clean-packages.sh
    clean-packages in the repository (used to clean out tons of old packages)
    "sudo ./clean-packages.sh /var/www/dmorris/pool/untangle/ 3 delete"
    deletes all but the newest three version of everything


gocr/HISTORY
History: (Changes,ChangeLog)

 0.41pre
       otsu.c concentrates now only on high contrast regions
       fix pnm reads for 2 byte pixels (--with-libpbm=no) 
       update man-page (mail me your suggestions)
       fix g++ warnings, float-OPs replaced by int-OPs
       spacing reviewed; make distance() more sensitive
       xml-objects (barcode, melted chars) now also handled with weights
       fix division by zero bug for vertical positioned characters
       default output is UTF8 now, UTF-encoding bug fixed
       added certainty option 
       added uninstall to Makefile
       debug image format changed to png (using pipe) or ppm (fall-back)
       much better word spacing (line-by-line based)
       better DOT_ABOVE recognition
       fix output of char groups or strings stored in database, utf8 input
       fix buffer overflow in barcode decode39
       fix lost comma on end of line
       internal vector format added for future use (faster, scalable, rotable)
       line detection extended
       internal list management rewritten to fix memory leaks and segfaults

 0.40
       update PNM file reader to maxval > 255
       (make rpm) updated
       barcode-patch UPC_addon by Michael van Rooyen
       CAPITAL_LETTER_A_WITH_OGONEK added
       no "(PICTURE)" output for UTF8+ASCII (better for Mobile OCR project)
       smooth_borders() bug fixed and reworked
       5x7 and prop10 font adaptions
       objects now detected by flood-fill algorithm (better?)
       XML-output changed
       changed auto dust detection (not final)

 0.39
       XML output added (subject of change, suggestions are welcome)
       netpbm-link-error fixed in gocr.c and configure.in:
         gocr.c: <config.h> changed to "config.h"
       configure-option --with-netpbm=PATH and --without-netpbm added
       update configure.in according to autoconf 2.57  
       wchar_h miss-configuration fixed in pgm2asc.c
       fix compiler warnings
       char filter accepts abbreviations now, like "0-9A-F" (but slow)
       update READMEde.txt
       output barcode tags (also improved recognition)
       fix pnm.c for files like example.eps.pbm
       fix detect.c for barcodes
       fix ocr0n.c 0<->8g

 0.38
       move UTF/HTML/TeX decoding to getTextLine, return (char *) now
       out_format HTML step towards detailed XML output
       correct line detection for footnotes (detect.c)
       "y" now seen as vowel (pgm2asc.c), I<vowel> susbtituted by l<vowel>
       &eacute;-detection, &aacute;-output fixed
       default dust_size is -1 now (auto detection = mean_size/10)
       char filter added
        ex: -C 0123456789ABCDEF    - recognize only hexcodes
       man page updated (hopefully correct syntax)
       database bug fixed (small fonts, example by Chris)
       several bugs fixed by W. Webber (thanks)
       speed improved by 3rd-pass matrix filter in pixel() (pixel.c) (code from W. Webber)
       bug in remove_dust (remove.c) fixed
       for fonts bigger than 20x40 smooth_borders() changed (b/w-scans)
       bug in O0-detection fixed

 0.37
        best-fit generates probability, not perfect but better results
        bug in line detection removed (happens for lot of small boxes)
        progress output (option -x <fileID|fname>)
        counting versions number as floating point now
        MACRON and DOT_ABOVE (not complete) defined (latin2)
        adaptions for 5x7 and 6x12 screen font
	doc/ocr.tex changed to doc/gocr.html (now independent of LaTeX)
        symbols {} added
        OCR-B font tested succesfull
        better headline/picture distinction
        bug removed (struct box.modifier is wchar_t now)

        known bugs: to much newlines

 0.3.6
        CARON and Omega defined,
        output of not defined chars (HTML="&#xxx;", TeX="\symbol{xxx}")
        system dependend bug: isupper(>255) SIGSEGV fixed
        better line detection for lines with lowercase chars only
        lot of possible SIGSEGV in list_del() fixed
        barcode recognition (UPC,code128)
        .ps .eps via pstopnm supported
        -m 256 switches off the main ocr engine (usefull together with -m 2 for identical chars)
        strings added to database ("ff","ft","special-symbol")
        gocr.tcl adapted to gocr v0.3
        internal detection probability introduced

 0.3.5
        minor and major fixes (string\0 bugs)
        memory leak fixes by Duncan Edwards
        layout analysis or zoning (-m 4) improved, 
        now it detects pictures and columns much better
        the behavior of setting threshold (-l) is slightly changed
        wcsdup defined for non-gnu-systems (BSD), further Problems?
        better context correction for 10 (IO,lO)
        Fixes for S.Koledin examples "GlS"
        Euro-currency-sign detection added
        better pitch estimation for proportional font (needs to be improved)
        make install DESTDIR= instead configure --prefix= (better?)
        use wchar_t by default, more simple code and -f works with nonLinuxOS
        line detection more robust against vertical glued chars (js)
        -f UTF8 added (usefull for xterm -u8), should be default?
        handle vertical glued boxes (ex: g over T)
 0.3.4
        some BSD adaptions (no WCHAR?), tell me if there are still problems
        use unicode in database (4-8 hex digits)
        new option: -p database_path/
        TILDE fixed, #, &AElig;, &Aring;, etc. added (swedish,norwegian)
        layout analysis improved
 0.3.3
        database (-m 2) bug fixed and interactive mode (-m 130) added
          its not finished, but you can test it
          result should be ok for machine generated images (no scans)           
        engine improved a bit
 0.3.2
        ocr-engine improved for screen fonts (thanks for examples)
        option -f [HTML,TeX,...] added
 0.3.1
        make install updated
 0.3.0  some parts of the code reviewed (most work done by Bruno Barberi Gnecco)

gocr/src/gocr.c
/*
This is a Optical-Character-Recognition program
Copyright (C) 2000-2006  Joerg Schulenburg

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

 see README for EMAIL-address

  sometimes I have written comments in german language, sorry for that

 This file was retrieved from pgm2asc.cc of Joerg, in order to have
 a library of the ocr-engine from Klaas Freitag
 
 */
#include "config.h"
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <string.h>
#ifdef HAVE_GETTIMEOFDAY 
#include <sys/time.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "pnm.h"
#include "pgm2asc.h"
#include "pcx.h"
#include "ocr0.h"		/* only_numbers */

/* the following line is used to name the package (make dist) */
#define version_string "0.41"

static void out_version(int v) {
  fprintf(stderr, " Optical Character Recognition --- gocr " version_string "\n");
  if (v)
    fprintf(stderr, " use option -h for help\n");
  if (v & 2)
    exit(1);
  return;
}

static void help(void) {
  out_version(0);
  /* output is shortened to essentials, see manual page for details */
  fprintf(stderr,
	  " using: gocr [options] pnm_file_name  # use - for stdin\n"
	  " options (see gocr manual pages for more details):\n"
	  " -h        - get this help\n"
	  " -i name   - input image file (pnm,pgm,pbm,ppm,pcx,...)\n"
	  " -o name   - output file  (redirection of stdout)\n"
	  " -e name   - logging file (redirection of stderr)\n"
	  " -x name   - progress output to fifo (see manual)\n"
	  " -p name   - database path including final slash (default is ./db/)\n"
	  " -f fmt    - output format (ISO8859_1 TeX HTML XML UTF8 ASCII)\n"
	  " -l num    - threshold grey level 0<160<=255 (0 = autodetect)\n"
	  " -d num    - dust_size (remove small clusters, -1 = autodetect)\n"
	  " -s num    - spacewidth/dots (0 = autodetect)\n"
	  " -v num    - verbose (see manual page)\n"
	  " -c string - list of chars (debugging, see manual)\n"
	  " -C string - char filter (ex. hexdigits: ""0-9A-Fx"", only ASCII)\n"
	  " -m num    - operation modes (bitpattern, see manual)\n"
	  " -a num      value of certainty (in percent, 0..100, default=95)\n"
	  " examples:\n"
	  "\tgocr -m 4 text1.pbm                   # do layout analyzis\n"
	  "\tgocr -m 130 -p ./database/ text1.pbm  # extend database\n"
	  "\tdjpeg -pnm -gray text.jpg | gocr -    # use jpeg-file via pipe\n"
	  "\n");
  exit(0);
}

#ifdef HAVE_GETTIMEOFDAY
/* from the glibc documentation */
static int timeval_subtract (struct timeval *result, struct timeval *x, 
    struct timeval *y) {

  /* Perform the carry for the later subtraction by updating Y. */
  if (x->tv_usec < y->tv_usec) {
    int nsec = (y->tv_usec - x->tv_usec) / 1000000 + 1;
    y->tv_usec -= 1000000 * nsec;
    y->tv_sec += nsec;
  }
  if (x->tv_usec - y->tv_usec > 1000000) {
    int nsec = (x->tv_usec - y->tv_usec) / 1000000;
    y->tv_usec += 1000000 * nsec;
    y->tv_sec -= nsec;
  }

  /* Compute the time remaining to wait.
     `tv_usec' is certainly positive. */
  result->tv_sec = x->tv_sec - y->tv_sec;
  result->tv_usec = x->tv_usec - y->tv_usec;

  /* Return 1 if result is negative. */
  return x->tv_sec < y->tv_sec;
}
#endif

static void process_arguments(job_t *job, int argn, char *argv[])
{
  int i;
  char *s1;

  assert(job);

  if (argn <= 1) {
    out_version(1);
    exit(0);
  }
#ifdef HAVE_PGM_H
  pnm_init(&argn, &argv);
#endif
  
  /* process arguments */
  for (i = 1; i < argn; i++) {
    if (strcmp(argv[i], "--help") == 0)
      help(); /* and quits */
    if (argv[i][0] == '-' && argv[i][1] != 0) {
      s1 = "";
      if (i + 1 < argn)
	s1 = argv[i + 1];
      switch (argv[i][1]) {
      case 'h': /* help */
	help();
	break;
      case 'i': /* input image file */
	job->src.fname = s1;
	i++;
	break;
      case 'e': /* logging file */
	if (s1[0] == '-' && s1[1] == '\0') {
#ifdef HAVE_UNISTD_H
          dup2(STDOUT_FILENO, STDERR_FILENO); /* -e /dev/stdout  works */
#else
	  fprintf(stderr, "stderr redirection not possible without unistd.h\n");
#endif           
	}
	else if (!freopen(s1, "w", stderr)) {
	  fprintf(stderr, "stderr redirection to %s failed\n", s1);
	}
	i++;
	break;
      case 'p': /* database path */
	job->cfg.db_path=s1;
	i++;
	break;
      case 'o': /* output file */
	if (s1[0] == '-' && s1[1] == '\0') {	/* default */
	}
	else if (!freopen(s1, "w", stdout)) {
	  fprintf(stderr, "stdout redirection to %s failed\n", s1);
	};
	i++;
	break;
      case 'f': /* output format */
        if (strcmp(s1, "ISO8859_1") == 0) job->cfg.out_format=ISO8859_1; else
        if (strcmp(s1, "TeX")       == 0) job->cfg.out_format=TeX; else
        if (strcmp(s1, "HTML")      == 0) job->cfg.out_format=HTML; else
        if (strcmp(s1, "XML")       == 0) job->cfg.out_format=XML; else
        if (strcmp(s1, "SGML")      == 0) job->cfg.out_format=SGML; else
        if (strcmp(s1, "UTF8")      == 0) job->cfg.out_format=UTF8; else
        if (strcmp(s1, "ASCII")     == 0) job->cfg.out_format=ASCII; else
        fprintf(stderr,"Warning: unknown format (-f %s)\n",s1);
        i++;
        break;
      case 'c': /* list of chars (_ = not recognized chars) */
	job->cfg.lc = s1;
	i++;
	break;
      case 'C': /* char filter, default: NULL (all chars) */
        /* ToDo: UTF8 input, wchar */
	job->cfg.cfilter = s1;
	i++;
	break;
      case 'd': /* dust size */
	job->cfg.dust_size = atoi(s1);
	i++;
	break;
      case 'l': /* grey level 0<160<=255, 0 for autodetect */
	job->cfg.cs = atoi(s1);
	i++;
	break;
      case 's': /* spacewidth/dots (0 = autodetect) */
	job->cfg.spc = atoi(s1);
	i++;
	break;
      case 'v': /* verbose mode */
	job->cfg.verbose |= atoi(s1);
	i++;
	break;
      case 'm': /* operation modes */
	job->cfg.mode |= atoi(s1);
	i++;
	break;
      case 'n': /* numbers only */
	job->cfg.only_numbers = atoi(s1);
	i++;
	break;
      case 'x': /* initialize progress output s1=fname */
	ini_progress(s1);
	i++;
	break;
      case 'a': /* set certainty */
	job->cfg.certainty = atoi(s1);;
	i++;
	break;
      default:
	fprintf(stderr, "# unknown option use -h for help\n");
      }
      continue;
    }
    else /* argument can be filename v0.2.5 */ if (argv[i][0] != '-'
						   || argv[i][1] == '\0' ) {
      job->src.fname = argv[i];
    }
  }
}

static void mark_start(job_t *job) {
  assert(job);

  if (job->cfg.verbose) {
    fprintf(stderr, "#");
    out_version(0);
    fprintf(stderr,
	    "# options are: -l %d -s %d -v %d -c %s -m %d -d %d -n %d %s\n",
	    job->cfg.cs, job->cfg.spc, job->cfg.verbose, job->cfg.lc, job->cfg.mode, 
            job->cfg.dust_size, job->cfg.only_numbers, job->src.fname);
#ifdef USE_UNICODE
    fprintf(stderr,"# using unicode\n");
#endif
#ifdef HAVE_GETTIMEOFDAY
    gettimeofday(&job->tmp.init_time, NULL);
#endif
  }
}

static void mark_end(job_t *job) {
  assert(job);

#ifdef HAVE_GETTIMEOFDAY
  /* show elapsed time */
  if (job->cfg.verbose) {
    struct timeval end, result;
    gettimeofday(&end, NULL);
    timeval_subtract(&result, &end, &job->tmp.init_time);
    fprintf(stderr,"Elapsed time: %d:%02d:%3.3f.\n", (int)result.tv_sec/60,
	(int)result.tv_sec%60, (float)result.tv_usec/1000);
  }
#endif
}

static void read_picture(job_t *job) {
  assert(job);

  if (strstr(job->src.fname, ".pcx"))
    readpcx(job->src.fname, &job->src.p, job->cfg.verbose);
  else
    readpgm(job->src.fname, &job->src.p, job->cfg.verbose);
}

// subject of change, we need more output for XML (ToDo)
void print_output(job_t *job) {
  int linecounter = 0;
  const char *line;

  assert(job);
        
  linecounter = 0;
  line = getTextLine(linecounter++);
  while (line) {
    // notice: decode() is shiftet to getTextLine since 0.38
    fputs(line, stdout);
    if (job->cfg.out_format==HTML) fputs("<br />",stdout);
    if (job->cfg.out_format!=XML)  fputc('\n', stdout);
    line = getTextLine(linecounter++);
  }
  free_textlines();
}

//FIXME jb: remove JOB;
job_t *JOB;


/* -------------------------------------------------------------
// ------   MAIN - replace this by your own aplication! 
// ------------------------------------------------------------- */
int main(int argn, char *argv[]) {
  job_t job;

  JOB = &job;
  setvbuf(stdout, (char *) NULL, _IONBF, 0);	/* not buffered */
  
  job_init(&job);
  
  process_arguments(&job, argn, argv);
  
  mark_start(&job);
 
  read_picture(&job);
  /* separation of main and rest for using as lib
     this will be changed later => introduction of set_option()
     for better communication to the engine  */

  /* call main loop */
  pgm2asc(&job);

  print_output(&job);
  
  mark_end(&job);
  
  job_free(&job);
  
  return 0;
}
